### PR TITLE
Fix bundle (re-)loading

### LINKED ISSUE #21758: Bundle extension loading but not working in 10.10.1
### Describe the Bug

My custom bundle extension seems to be loading and visible in the Settings -> Extensions tab (see image).
Screenshot 2024-03-08 at 09 47 05

But they are not displayed in the collection item forms and are show up as "...not found" in the field interface settings (see image bellow).
Screenshot 2024-03-08 at 09 49 50

Also, the endpoints and hooks don't seem to be working. 

I have made sure my bundle meets all of the latests requirements  explained here https://docs.directus.io/releases/breaking-changes.html#deprecated-typed-extension-folders. The bundle is currently being imported as a npm package (not placed in the extensions folder).

Also one more thing to note is that when I check the "/extensions/sources/index.js" being loading from the app it seems to be completely empty (see image).

Screenshot 2024-03-08 at 09 58 29

### To Reproduce

Wouldn't know exactly how to reproduce the issue with other extensions. I have a bundle type extension in typescript with many interfaces, endpoints, hooks and some displays and not sure if one of these is causing the issue. Might also be how the npm package is being build.

### Directus Version

v10.10.1

### Hosting Strategy

Self-Hosted (Custom)

### LINKED ISSUE #21922: New extensions in bundle aren't reloaded in development mode
### Describe the Bug

Whenever I edit my extension's package.json while developing, directus does not reload it. Even een reboot does not solve my issue. I have to drop my database and rebootstrap.

### To Reproduce

1. Create a directus extension folder in `extension/demo` with the "bundle" type.
2. Add a hooks extension to the bundle called "hooks1"
3. Bootstrap and run directus with `EXTENSIONS_AUTO_RELOAD=true`
4. Add another hooks extension to the bundle called "hooks2"
5. The demo/hooks2 extension is not loaded
6. Restart directus
7. The demo/hooks2 extention is not loaded
8. Drop database and rebootstrap
9. The demo/hooks2 extension is loaded

### Directus Version

v10.10.4

### Hosting Strategy

Self-Hosted (Custom)

### LINKED ISSUE #21942: Extensions: Changing name of an `entry` on a `bundle`  it's not reflected
### Describe the Bug

While working with a `bundle` extension, I have tried to change the name of an `endpoint`.
I have done this by change the name of the `entry` in the `package.json`.
As result I was expecting to access the custom endpoint with the new name, but it was only accessible using the old name.

### To Reproduce

- Create a folder
- Run `npx create-directus-extension@latest` and choose bundle
- Enter newly created folder
- Run `npm run add` and choose `endpoint` with name "lorem"
- Access `/lorem` and you should see "Hello, World!"
- Copy files to extensions folder
- Edit `package.json` and change name "lorem" to "ipsum"
- Restart Directus
- Access `/ipsum` and it will send 404. Access `/lorem` and you should see "Hello, World!"

### Directus Version

10.10.4 (main)

### Hosting Strategy

Self-Hosted (Custom)

### LINKED ISSUE #21946: Extensions: Adding a new entry to bundle does not update extensions list
### Describe the Bug

After adding a new entry to the bundle, in this case a display, this new entry does not appear on extensions list.
Although the entry is usable, at least the ones of type display.

### To Reproduce

- Create a folder
- Run `npx create-directus-extension@latest` and choose bundle
- Enter newly created folder
- Run `npm run add` and choose `display` with name "lorem"
- Copy files to extensions folder
- Run `npm run add` and choose `display` with name "ipsum"
- Go to Extensions and check that "ipsum" display does not appear there

### Directus Version

10.10.4 (main)

### Hosting Strategy

Self-Hosted (Custom)

### PR DESCRIPTION
**Pull request by hanneskuettner**
## Scope

What's changed:

- Instead of skipping over already registered bundle extensions, update the `directus_extensions` bundle entry records when loading a bundle extension, by removing stale ones and adding new ones
- Instead of using `database.insert` I'm now using the `extensionsItemService.insert`, was there any reason it was not used before?

## Potential Risks / Drawbacks

- This approach removes bundle extension entries that are no longer present, potentially losing the extension enabled state. In practice this should not matter, however.

## Review Notes / Questions

- I would like Pascal to review this since he initially wrote this and might have some thoughts on it.

---

Fixes #21942, fixes #21946, fixes #21922

- [x] Should also fix #21758, to be checked

### CODE REVIEW

File: .changeset/hot-dolphins-shop.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@directus/api': patch
+---
+
+Fix bundle extension loading by updating extension settings table
```
Comment (paescuj) [inline .changeset/hot-dolphins-shop.md:5]:
```suggestion
Fixed bundle extension loading by updating extension settings table
```

File: api/src/extensions/lib/get-extensions-settings.ts [hunk 2]
```diff
@@ -28,11 +28,44 @@ export const getExtensionsSettings = async ({
...
+	const removedSettingIds: string[] = [];
 
 	const localSettings = existingSettings.filter(({ source }) => source === 'local');
 	const registrySettings = existingSettings.filter(({ source }) => source === 'registry');
 	const moduleSettings = existingSettings.filter(({ source }) => source === 'module');
 
+	const updateBundleEntriesSettings = (
+		extensionSettings: ExtensionSettings,
+		extension: Extension,
+		settings: ExtensionSettings[],
```
Comment (paescuj) [inline api/src/extensions/lib/get-extensions-settings.ts:40]: Would be awesome if we can find more expressive names for the params 🙏 E.g.
```suggestion
		bundle: BundleExtension,
		bundleSettings: ExtensionSettings,
		allSettings: ExtensionSettings[],
```

File: api/src/extensions/lib/get-extensions-settings.ts [hunk 3]
```diff
@@ -66,8 +99,12 @@ export const getExtensionsSettings = async ({
 	};
 
 	for (const [folder, extension] of local.entries()) {
-		const settingsExist = localSettings.some((settings) => settings.folder === folder);
-		if (settingsExist) continue;
+		const existingSettings = localSettings.find((settings) => settings.folder === folder);
+
+		if (existingSettings) {
```
Comment (paescuj) [inline api/src/extensions/lib/get-extensions-settings.ts:104]: Place check here (as it's done for module and registry extensions)

```suggestion
		if (existingSettings) {
			if (extension.type === 'bundle') {
				updateBundleEntriesSettings(existingSettings, extension, localSettings);
			}

```

File: api/src/extensions/lib/get-extensions-settings.ts [hunk 4]
```diff
@@ -28,11 +28,44 @@ export const getExtensionsSettings = async ({
...
 	const registrySettings = existingSettings.filter(({ source }) => source === 'registry');
 	const moduleSettings = existingSettings.filter(({ source }) => source === 'module');
 
+	const updateBundleEntriesSettings = (
+		extensionSettings: ExtensionSettings,
+		extension: Extension,
+		settings: ExtensionSettings[],
+	) => {
+		if (extension.type !== 'bundle') return;
+
```
Comment (paescuj) [inline api/src/extensions/lib/get-extensions-settings.ts:43]: Let's move this check outside of this function (and expect only `BundleExtension` to be passed)

```suggestion
```

**Review by paescuj — COMMENTED** [review]
As already discussed, I think we can consider #21758 fixed with this PR as the missing bundle entries should be added now.

**Review by br41nslug — CHANGES_REQUESTED** [review]
- [x] renaming an endpoint in a bundle (#21942)

* ~~successfully updates the database but doesnt actually load the newly named endpoint either on auto-reload or on a manual restart~~ (endpoint names get compiled into the bundle unlike regular endpoints, so missed a rebuild in that specific test)


- [x] adding a new entry to a bundle (#21946, #21922) (tested with `displays`, `endpoints` and `hooks`)
- [x] removing an entry from a bundle

**Review by paescuj — APPROVED** [review]
Great work 🎉

**Review by br41nslug — APPROVED** [review]

**Pull request merged**